File: examples/cpp/tests/player_steps.cpp
// GTest must be included before cucumber-cpp autodetect for framework detection
#include <gtest/gtest.h>
#include <cucumber-cpp/autodetect.hpp>

#include "angzarr/errors.hpp"
#include "deposit_handler.hpp"
#include "examples/player.pb.h"
#include "examples/poker_types.pb.h"
#include "player_state.hpp"
#include "register_handler.hpp"
#include "release_handler.hpp"
#include "reserve_handler.hpp"
#include "test_context.hpp"
#include "withdraw_handler.hpp"

using cucumber::ScenarioScope;

namespace {

/// Helper to create a table root ID from a string.
std::string make_table_root(const std::string& table_id) {
    // In production, this would be a proper UUID. For tests, use the string as-is.
    return table_id;
}

}  // anonymous namespace

// ==========================================================================
// Given Steps - Setting up state from events
// ==========================================================================

GIVEN("^no prior events for the player aggregate$") {
    tests::g_context.event_pages.clear();
    tests::g_context.player_state = player::PlayerState{};
}

GIVEN("^a PlayerRegistered event for \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, name);

    examples::PlayerRegistered event;
    event.set_display_name(name);
    event.set_email(name + "@example.com");
    event.set_player_type(examples::PlayerType::HUMAN);

    tests::g_context.add_event(event);
    tests::g_context.rebuild_player_state();
}

GIVEN("^a FundsDeposited event with amount (\\d+)$") {
    REGEX_PARAM(int64_t, amount);

    // Calculate new balance based on current state
    int64_t new_balance = tests::g_context.player_state.bankroll + amount;

    examples::FundsDeposited event;
    event.mutable_amount()->set_amount(amount);
    event.mutable_amount()->set_currency_code("CHIPS");
    event.mutable_new_balance()->set_amount(new_balance);
    event.mutable_new_balance()->set_currency_code("CHIPS");

    tests::g_context.add_event(event);
    tests::g_context.rebuild_player_state();
}

GIVEN("^a FundsReserved event with amount (\\d+) for table \"([^\"]*)\"$") {
    REGEX_PARAM(int64_t, amount);
    REGEX_PARAM(std::string, table_id);

    int64_t new_reserved = tests::g_context.player_state.reserved_funds + amount;

    examples::FundsReserved event;
    event.set_table_root(make_table_root(table_id));
    event.mutable_amount()->set_amount(amount);
    event.mutable_amount()->set_currency_code("CHIPS");
    event.mutable_new_reserved_balance()->set_amount(new_reserved);
    event.mutable_new_reserved_balance()->set_currency_code("CHIPS");
    event.mutable_new_available_balance()->set_amount(
        tests::g_context.player_state.bankroll - new_reserved);
    event.mutable_new_available_balance()->set_currency_code("CHIPS");

    tests::g_context.add_event(event);
    tests::g_context.rebuild_player_state();
}

// ==========================================================================
// When Steps - Handling commands
// ==========================================================================

WHEN("^I handle a RegisterPlayer command with name \"([^\"]*)\" and email \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, name);
    REGEX_PARAM(std::string, email);

    tests::g_context.clear_error();

    examples::RegisterPlayer cmd;
    cmd.set_display_name(name);
    cmd.set_email(email);
    cmd.set_player_type(examples::PlayerType::HUMAN);

    try {
        auto event = player::handlers::handle_register(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I handle a RegisterPlayer command with name \"([^\"]*)\" and email \"([^\"]*)\" as AI$") {
    REGEX_PARAM(std::string, name);
    REGEX_PARAM(std::string, email);

    tests::g_context.clear_error();

    examples::RegisterPlayer cmd;
    cmd.set_display_name(name);
    cmd.set_email(email);
    cmd.set_player_type(examples::PlayerType::AI);
    cmd.set_ai_model_id("gpt-4");

    try {
        auto event = player::handlers::handle_register(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I handle a DepositFunds command with amount (\\d+)$") {
    REGEX_PARAM(int64_t, amount);

    tests::g_context.clear_error();

    examples::DepositFunds cmd;
    cmd.mutable_amount()->set_amount(amount);
    cmd.mutable_amount()->set_currency_code("CHIPS");

    try {
        auto event = player::handlers::handle_deposit(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I handle a WithdrawFunds command with amount (\\d+)$") {
    REGEX_PARAM(int64_t, amount);

    tests::g_context.clear_error();

    examples::WithdrawFunds cmd;
    cmd.mutable_amount()->set_amount(amount);
    cmd.mutable_amount()->set_currency_code("CHIPS");

    try {
        auto event = player::handlers::handle_withdraw(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I handle a ReserveFunds command with amount (\\d+) for table \"([^\"]*)\"$") {
    REGEX_PARAM(int64_t, amount);
    REGEX_PARAM(std::string, table_id);

    tests::g_context.clear_error();

    examples::ReserveFunds cmd;
    cmd.set_table_root(make_table_root(table_id));
    cmd.mutable_amount()->set_amount(amount);
    cmd.mutable_amount()->set_currency_code("CHIPS");

    try {
        auto event = player::handlers::handle_reserve(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I handle a ReleaseFunds command for table \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, table_id);

    tests::g_context.clear_error();

    examples::ReleaseFunds cmd;
    cmd.set_table_root(make_table_root(table_id));

    try {
        auto event = player::handlers::handle_release(cmd, tests::g_context.player_state);
        tests::g_context.set_result(event);
    } catch (const angzarr::CommandRejectedError& e) {
        tests::g_context.set_error(e.what(), e.status_code);
    }
}

WHEN("^I rebuild the player state$") { tests::g_context.rebuild_player_state(); }

// ==========================================================================
// Then Steps - Assertions on results
// ==========================================================================

THEN("^the result is a examples\\.PlayerRegistered event$") {
    ASSERT_TRUE(tests::g_context.result_event.has_value()) << "Expected a result event";
    auto event = tests::g_context.get_result_as<examples::PlayerRegistered>();
    ASSERT_TRUE(event.has_value()) << "Expected PlayerRegistered event";
}

THEN("^the result is a examples\\.FundsDeposited event$") {
    ASSERT_TRUE(tests::g_context.result_event.has_value()) << "Expected a result event";
    auto event = tests::g_context.get_result_as<examples::FundsDeposited>();
    ASSERT_TRUE(event.has_value()) << "Expected FundsDeposited event";
}

THEN("^the result is a examples\\.FundsWithdrawn event$") {
    ASSERT_TRUE(tests::g_context.result_event.has_value()) << "Expected a result event";
    auto event = tests::g_context.get_result_as<examples::FundsWithdrawn>();
    ASSERT_TRUE(event.has_value()) << "Expected FundsWithdrawn event";
}

THEN("^the result is a examples\\.FundsReserved event$") {
    ASSERT_TRUE(tests::g_context.result_event.has_value()) << "Expected a result event";
    auto event = tests::g_context.get_result_as<examples::FundsReserved>();
    ASSERT_TRUE(event.has_value()) << "Expected FundsReserved event";
}

THEN("^the result is a examples\\.FundsReleased event$") {
    ASSERT_TRUE(tests::g_context.result_event.has_value()) << "Expected a result event";
    auto event = tests::g_context.get_result_as<examples::FundsReleased>();
    ASSERT_TRUE(event.has_value()) << "Expected FundsReleased event";
}

THEN("^the player event has display_name \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, expected_name);

    auto event = tests::g_context.get_result_as<examples::PlayerRegistered>();
    ASSERT_TRUE(event.has_value()) << "Expected PlayerRegistered event";
    ASSERT_EQ(event->display_name(), expected_name);
}

THEN("^the player event has player_type \"([^\"]*)\"$") {
    REGEX_PARAM(std::string, expected_type);

    auto event = tests::g_context.get_result_as<examples::PlayerRegistered>();
    ASSERT_TRUE(event.has_value()) << "Expected PlayerRegistered event";

    examples::PlayerType expected;
    if (expected_type == "HUMAN") {
        expected = examples::PlayerType::HUMAN;
    } else if (expected_type == "AI") {
        expected = examples::PlayerType::AI;
    } else {
        expected = examples::PlayerType::PLAYER_TYPE_UNSPECIFIED;
    }
    ASSERT_EQ(event->player_type(), expected);
}

THEN("^the player event has amount (\\d+)$") {
    REGEX_PARAM(int64_t, expected_amount);

    // This could be any event with an amount field - check each type
    if (auto event = tests::g_context.get_result_as<examples::FundsDeposited>()) {
        ASSERT_EQ(event->amount().amount(), expected_amount);
    } else if (auto event = tests::g_context.get_result_as<examples::FundsWithdrawn>()) {
        ASSERT_EQ(event->amount().amount(), expected_amount);
    } else if (auto event = tests::g_context.get_result_as<examples::FundsReserved>()) {
        ASSERT_EQ(event->amount().amount(), expected_amount);
    } else if (auto event = tests::g_context.get_result_as<examples::FundsReleased>()) {
        ASSERT_EQ(event->amount().amount(), expected_amount);
    } else {
        FAIL() << "No event with 'amount' field found";
    }
}

THEN("^the player event has new_balance (\\d+)$") {
    REGEX_PARAM(int64_t, expected_balance);

    if (auto event = tests::g_context.get_result_as<examples::FundsDeposited>()) {
        ASSERT_EQ(event->new_balance().amount(), expected_balance);
    } else if (auto event = tests::g_context.get_result_as<examples::FundsWithdrawn>()) {
        ASSERT_EQ(event->new_balance().amount(), expected_balance);
    } else {
        FAIL() << "No event with 'new_balance' field found";
    }
}

THEN("^the player event has new_available_balance (\\d+)$") {
    REGEX_PARAM(int64_t, expected_balance);

    if (auto event = tests::g_context.get_result_as<examples::FundsReserved>()) {
        ASSERT_EQ(event->new_available_balance().amount(), expected_balance);
    } else if (auto event = tests::g_context.get_result_as<examples::FundsReleased>()) {
        ASSERT_EQ(event->new_available_balance().amount(), expected_balance);
    } else {
        FAIL() << "No event with 'new_available_balance' field found";
    }
}

// State assertions
THEN("^the player state has bankroll (\\d+)$") {
    REGEX_PARAM(int64_t, expected);
    ASSERT_EQ(tests::g_context.player_state.bankroll, expected);
}

THEN("^the player state has reserved_funds (\\d+)$") {
    REGEX_PARAM(int64_t, expected);
    ASSERT_EQ(tests::g_context.player_state.reserved_funds, expected);
}

THEN("^the player state has available_balance (\\d+)$") {
    REGEX_PARAM(int64_t, expected);
    ASSERT_EQ(tests::g_context.player_state.available_balance(), expected);
}

File: examples/go/tests/player_steps.go
package tests

import (
	"context"
	"encoding/hex"
	"fmt"
	"strings"

	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"github.com/benjaminabbitt/angzarr/examples/go/player/agg/handlers"
	"github.com/cucumber/godog"
	"github.com/google/uuid"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// PlayerContext holds state for player aggregate scenarios
type PlayerContext struct {
	eventPages  []*pb.EventPage
	state       handlers.PlayerState
	resultEvent *anypb.Any
	resultBook  *pb.EventBook
	lastError   error
}

func newPlayerContext() *PlayerContext {
	return &PlayerContext{
		eventPages: []*pb.EventPage{},
		state:      handlers.NewPlayerState(),
	}
}

// InitPlayerSteps registers player aggregate step definitions
func InitPlayerSteps(ctx *godog.ScenarioContext) {
	pc := newPlayerContext()

	// Reset before each scenario
	ctx.Before(func(ctx context.Context, sc *godog.Scenario) (context.Context, error) {
		pc.eventPages = []*pb.EventPage{}
		pc.state = handlers.NewPlayerState()
		pc.resultEvent = nil
		pc.resultBook = nil
		pc.lastError = nil
		return ctx, nil
	})

	// Given steps
	ctx.Step(`^no prior events for the player aggregate$`, pc.noPriorEvents)
	ctx.Step(`^a PlayerRegistered event for "([^"]*)"$`, pc.playerRegisteredFor)
	ctx.Step(`^a FundsDeposited event with amount (\d+)$`, pc.fundsDepositedWithAmount)
	ctx.Step(`^a FundsReserved event with amount (\d+) for table "([^"]*)"$`, pc.fundsReservedForTable)

	// When steps
	ctx.Step(`^I handle a RegisterPlayer command with name "([^"]*)" and email "([^"]*)"$`, pc.handleRegisterPlayer)
	ctx.Step(`^I handle a RegisterPlayer command with name "([^"]*)" and email "([^"]*)" as AI$`, pc.handleRegisterPlayerAI)
	ctx.Step(`^I handle a DepositFunds command with amount (\d+)$`, pc.handleDepositFunds)
	ctx.Step(`^I handle a WithdrawFunds command with amount (\d+)$`, pc.handleWithdrawFunds)
	ctx.Step(`^I handle a ReserveFunds command with amount (\d+) for table "([^"]*)"$`, pc.handleReserveFunds)
	ctx.Step(`^I handle a ReleaseFunds command for table "([^"]*)"$`, pc.handleReleaseFunds)
	ctx.Step(`^I rebuild the player state$`, pc.rebuildPlayerState)

	// Then steps
	ctx.Step(`^the result is a (?:examples\.)?PlayerRegistered event$`, pc.resultIsPlayerRegistered)
	ctx.Step(`^the result is a (?:examples\.)?FundsDeposited event$`, pc.resultIsFundsDeposited)
	ctx.Step(`^the result is a (?:examples\.)?FundsWithdrawn event$`, pc.resultIsFundsWithdrawn)
	ctx.Step(`^the result is a (?:examples\.)?FundsReserved event$`, pc.resultIsFundsReserved)
	ctx.Step(`^the result is a (?:examples\.)?FundsReleased event$`, pc.resultIsFundsReleased)
	ctx.Step(`^the player event has display_name "([^"]*)"$`, pc.eventHasDisplayName)
	ctx.Step(`^the player event has player_type "([^"]*)"$`, pc.eventHasPlayerType)
	ctx.Step(`^the player event has amount (\d+)$`, pc.eventHasAmount)
	ctx.Step(`^the player event has new_balance (\d+)$`, pc.eventHasNewBalance)
	ctx.Step(`^the player event has new_available_balance (\d+)$`, pc.eventHasNewAvailableBalance)
	ctx.Step(`^the player state has bankroll (\d+)$`, pc.stateHasBankroll)
	ctx.Step(`^the player state has reserved_funds (\d+)$`, pc.stateHasReservedFunds)
	ctx.Step(`^the player state has available_balance (\d+)$`, pc.stateHasAvailableBalance)
	// Note: "command fails with status" and "error message contains" are registered in common_steps.go
}

// Helper functions

func (pc *PlayerContext) makeEventPage(event *anypb.Any) *pb.EventPage {
	return &pb.EventPage{
		Header:    &pb.PageHeader{SequenceType: &pb.PageHeader_Sequence{Sequence: uint32(len(pc.eventPages))}},
		CreatedAt: timestamppb.Now(),
		Payload:   &pb.EventPage_Event{Event: event},
	}
}

func (pc *PlayerContext) addEvent(event *anypb.Any) {
	pc.eventPages = append(pc.eventPages, pc.makeEventPage(event))
	pc.rebuildState()
}

func (pc *PlayerContext) rebuildState() {
	id := uuid.New()
	eventBook := &pb.EventBook{
		Cover: &pb.Cover{
			Domain: "player",
			Root:   &pb.UUID{Value: id[:]},
		},
		Pages:        pc.eventPages,
		NextSequence: uint32(len(pc.eventPages)),
	}
	pc.state = handlers.RebuildState(eventBook)
}

func (pc *PlayerContext) makeCommandBook() *pb.CommandBook {
	id := uuid.New()
	return &pb.CommandBook{
		Cover: &pb.Cover{
			Domain: "player",
			Root:   &pb.UUID{Value: id[:]},
		},
	}
}

// Given step implementations

func (pc *PlayerContext) noPriorEvents() error {
	pc.eventPages = []*pb.EventPage{}
	pc.state = handlers.NewPlayerState()
	return nil
}

func (pc *PlayerContext) playerRegisteredFor(name string) error {
	event := &examples.PlayerRegistered{
		DisplayName:  name,
		Email:        strings.ToLower(name) + "@example.com",
		PlayerType:   examples.PlayerType_HUMAN,
		RegisteredAt: timestamppb.Now(),
	}
	eventAny, err := anypb.New(event)
	if err != nil {
		return err
	}
	pc.addEvent(eventAny)
	return nil
}

func (pc *PlayerContext) fundsDepositedWithAmount(amount int) error {
	newBalance := pc.state.Bankroll + int64(amount)
	event := &examples.FundsDeposited{
		Amount:      &examples.Currency{Amount: int64(amount), CurrencyCode: "CHIPS"},
		NewBalance:  &examples.Currency{Amount: newBalance, CurrencyCode: "CHIPS"},
		DepositedAt: timestamppb.Now(),
	}
	eventAny, err := anypb.New(event)
	if err != nil {
		return err
	}
	pc.addEvent(eventAny)
	return nil
}

func (pc *PlayerContext) fundsReservedForTable(amount int, tableID string) error {
	newReserved := pc.state.ReservedFunds + int64(amount)
	newAvailable := pc.state.Bankroll - newReserved
	event := &examples.FundsReserved{
		Amount:              &examples.Currency{Amount: int64(amount), CurrencyCode: "CHIPS"},
		TableRoot:           []byte(tableID),
		NewAvailableBalance: &examples.Currency{Amount: newAvailable, CurrencyCode: "CHIPS"},
		NewReservedBalance:  &examples.Currency{Amount: newReserved, CurrencyCode: "CHIPS"},
		ReservedAt:          timestamppb.Now(),
	}
	eventAny, err := anypb.New(event)
	if err != nil {
		return err
	}
	pc.addEvent(eventAny)
	return nil
}

// When step implementations

func (pc *PlayerContext) handleRegisterPlayer(name, email string) error {
	cmd := &examples.RegisterPlayer{
		DisplayName: name,
		Email:       email,
		PlayerType:  examples.PlayerType_HUMAN,
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleRegisterPlayer(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) handleRegisterPlayerAI(name, email string) error {
	cmd := &examples.RegisterPlayer{
		DisplayName: name,
		Email:       email,
		PlayerType:  examples.PlayerType_AI,
		AiModelId:   "gpt-4",
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleRegisterPlayer(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) handleDepositFunds(amount int) error {
	cmd := &examples.DepositFunds{
		Amount: &examples.Currency{Amount: int64(amount), CurrencyCode: "CHIPS"},
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleDepositFunds(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) handleWithdrawFunds(amount int) error {
	cmd := &examples.WithdrawFunds{
		Amount: &examples.Currency{Amount: int64(amount), CurrencyCode: "CHIPS"},
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleWithdrawFunds(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) handleReserveFunds(amount int, tableID string) error {
	cmd := &examples.ReserveFunds{
		Amount:    &examples.Currency{Amount: int64(amount), CurrencyCode: "CHIPS"},
		TableRoot: []byte(tableID),
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleReserveFunds(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) handleReleaseFunds(tableID string) error {
	cmd := &examples.ReleaseFunds{
		TableRoot: []byte(tableID),
	}
	cmdAny, err := anypb.New(cmd)
	if err != nil {
		return err
	}

	result, err := handlers.HandleReleaseFunds(pc.makeCommandBook(), cmdAny, pc.state, uint32(len(pc.eventPages)))
	pc.lastError = err
	SetLastError(pc.lastError)
	if err == nil && result != nil && len(result.Pages) > 0 {
		pc.resultBook = result
		if event, ok := result.Pages[0].Payload.(*pb.EventPage_Event); ok {
			pc.resultEvent = event.Event
		}
	}
	return nil
}

func (pc *PlayerContext) rebuildPlayerState() error {
	pc.rebuildState()
	return nil
}

// Then step implementations

func (pc *PlayerContext) resultIsPlayerRegistered() error {
	if pc.lastError != nil {
		return fmt.Errorf("expected success but got error: %v", pc.lastError)
	}
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	if !pc.resultEvent.MessageIs(&examples.PlayerRegistered{}) {
		return fmt.Errorf("expected PlayerRegistered event, got %s", pc.resultEvent.TypeUrl)
	}
	return nil
}

func (pc *PlayerContext) resultIsFundsDeposited() error {
	if pc.lastError != nil {
		return fmt.Errorf("expected success but got error: %v", pc.lastError)
	}
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	if !pc.resultEvent.MessageIs(&examples.FundsDeposited{}) {
		return fmt.Errorf("expected FundsDeposited event, got %s", pc.resultEvent.TypeUrl)
	}
	return nil
}

func (pc *PlayerContext) resultIsFundsWithdrawn() error {
	if pc.lastError != nil {
		return fmt.Errorf("expected success but got error: %v", pc.lastError)
	}
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	if !pc.resultEvent.MessageIs(&examples.FundsWithdrawn{}) {
		return fmt.Errorf("expected FundsWithdrawn event, got %s", pc.resultEvent.TypeUrl)
	}
	return nil
}

func (pc *PlayerContext) resultIsFundsReserved() error {
	if pc.lastError != nil {
		return fmt.Errorf("expected success but got error: %v", pc.lastError)
	}
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	if !pc.resultEvent.MessageIs(&examples.FundsReserved{}) {
		return fmt.Errorf("expected FundsReserved event, got %s", pc.resultEvent.TypeUrl)
	}
	return nil
}

func (pc *PlayerContext) resultIsFundsReleased() error {
	if pc.lastError != nil {
		return fmt.Errorf("expected success but got error: %v", pc.lastError)
	}
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	if !pc.resultEvent.MessageIs(&examples.FundsReleased{}) {
		return fmt.Errorf("expected FundsReleased event, got %s", pc.resultEvent.TypeUrl)
	}
	return nil
}

func (pc *PlayerContext) eventHasDisplayName(name string) error {
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	var event examples.PlayerRegistered
	if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
		return err
	}
	if event.DisplayName != name {
		return fmt.Errorf("expected display_name=%s, got %s", name, event.DisplayName)
	}
	return nil
}

func (pc *PlayerContext) eventHasPlayerType(ptype string) error {
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}
	var event examples.PlayerRegistered
	if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
		return err
	}
	expected := examples.PlayerType(examples.PlayerType_value[ptype])
	if event.PlayerType != expected {
		return fmt.Errorf("expected player_type=%s, got %s", ptype, event.PlayerType.String())
	}
	return nil
}

func (pc *PlayerContext) eventHasAmount(amount int) error {
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}

	// Try different event types
	if pc.resultEvent.MessageIs(&examples.FundsDeposited{}) {
		var event examples.FundsDeposited
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.Amount.Amount != int64(amount) {
			return fmt.Errorf("expected amount=%d, got %d", amount, event.Amount.Amount)
		}
		return nil
	}
	if pc.resultEvent.MessageIs(&examples.FundsWithdrawn{}) {
		var event examples.FundsWithdrawn
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.Amount.Amount != int64(amount) {
			return fmt.Errorf("expected amount=%d, got %d", amount, event.Amount.Amount)
		}
		return nil
	}
	if pc.resultEvent.MessageIs(&examples.FundsReserved{}) {
		var event examples.FundsReserved
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.Amount.Amount != int64(amount) {
			return fmt.Errorf("expected amount=%d, got %d", amount, event.Amount.Amount)
		}
		return nil
	}
	if pc.resultEvent.MessageIs(&examples.FundsReleased{}) {
		var event examples.FundsReleased
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.Amount.Amount != int64(amount) {
			return fmt.Errorf("expected amount=%d, got %d", amount, event.Amount.Amount)
		}
		return nil
	}

	return fmt.Errorf("unknown event type: %s", pc.resultEvent.TypeUrl)
}

func (pc *PlayerContext) eventHasNewBalance(balance int) error {
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}

	if pc.resultEvent.MessageIs(&examples.FundsDeposited{}) {
		var event examples.FundsDeposited
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.NewBalance.Amount != int64(balance) {
			return fmt.Errorf("expected new_balance=%d, got %d", balance, event.NewBalance.Amount)
		}
		return nil
	}
	if pc.resultEvent.MessageIs(&examples.FundsWithdrawn{}) {
		var event examples.FundsWithdrawn
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.NewBalance.Amount != int64(balance) {
			return fmt.Errorf("expected new_balance=%d, got %d", balance, event.NewBalance.Amount)
		}
		return nil
	}

	return fmt.Errorf("event type %s does not have new_balance", pc.resultEvent.TypeUrl)
}

func (pc *PlayerContext) eventHasNewAvailableBalance(balance int) error {
	if pc.resultEvent == nil {
		return fmt.Errorf("no result event")
	}

	if pc.resultEvent.MessageIs(&examples.FundsReserved{}) {
		var event examples.FundsReserved
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.NewAvailableBalance.Amount != int64(balance) {
			return fmt.Errorf("expected new_available_balance=%d, got %d", balance, event.NewAvailableBalance.Amount)
		}
		return nil
	}
	if pc.resultEvent.MessageIs(&examples.FundsReleased{}) {
		var event examples.FundsReleased
		if err := pc.resultEvent.UnmarshalTo(&event); err != nil {
			return err
		}
		if event.NewAvailableBalance.Amount != int64(balance) {
			return fmt.Errorf("expected new_available_balance=%d, got %d", balance, event.NewAvailableBalance.Amount)
		}
		return nil
	}

	return fmt.Errorf("event type %s does not have new_available_balance", pc.resultEvent.TypeUrl)
}

func (pc *PlayerContext) stateHasBankroll(amount int) error {
	if pc.state.Bankroll != int64(amount) {
		return fmt.Errorf("expected bankroll=%d, got %d", amount, pc.state.Bankroll)
	}
	return nil
}

func (pc *PlayerContext) stateHasReservedFunds(amount int) error {
	if pc.state.ReservedFunds != int64(amount) {
		return fmt.Errorf("expected reserved_funds=%d, got %d", amount, pc.state.ReservedFunds)
	}
	return nil
}

func (pc *PlayerContext) stateHasAvailableBalance(amount int) error {
	available := pc.state.AvailableBalance()
	if available != int64(amount) {
		return fmt.Errorf("expected available_balance=%d, got %d", amount, available)
	}
	return nil
}

// Ensure TableReservations key uses hex encoding consistent with handlers
func init() {
	_ = hex.EncodeToString([]byte{})
}
